### DISCUSSION TITLE
Why does Fixture.Freeze return non-default values when using with AutoMoqCustomization and Moq?

### DISCUSSION THREAD

**Discussion by ivansorokinf** [AutoFixture/AutoFixture#1312] [Q&A]
Moq: 4.16.1
Autofixture: 4.17.0
Autofixture.AutoMoq: 4.17.0
.NET: .Net Framework 4.8

Problem: `fixture.Freeze()` setups some return values instead of default (null for references etc) for methods which is not a default Moq behaviour. I figured out that the reason of it lays in `DefaultValue=Mock` of Mocks created by Freeze method. If I override it after freezing with `DefaultValue=Empty`, then it works as expected. Is it any way to configure my fixture to use this logic globally? And why is it configured in fixture with `DefaultValue=Mock` in the first place?

```cs
public class Dummy
{
    public string DummyA { get; set; }
    public string DummyB { get; set; }
}

public interface IDep
{
    Dummy Get(int a);
}

public class MyClass
{
    private readonly IDep dep;

    public MyClass(IDep dep)
    {
        this.dep = dep;
    }

    public Dummy Method(int a) => dep.Get(a);
}

public class Tests
{
    [Test]
    public void MyTestCase1()
    {
        var fixture = new Fixture();
        fixture.Customize(new AutoMoqCustomization());
        
        var dep = fixture.Freeze<Mock<IDep>>();
        var myClass = fixture.Create<MyClass>();
        
        Assert.That(dep.Object.Get(5), Is.Null); //false
        Assert.That(myClass.Method(5), Is.Null); //false
    }
        
    [Test]
    public void MyTestCase2()
    {
        var fixture = new Fixture();
        fixture.Customize(new AutoMoqCustomization());

        var dep = new Mock<IDep>();
        fixture.Inject(dep.Object);
        var myClass = fixture.Create<MyClass>();
        
        Assert.That(dep.Object.Get(5), Is.Null); //true
        Assert.That(myClass.Method(5), Is.Null); //true
    }
    
    [Test]
    public void MyTestCase3()
    {
        var fixture = new Fixture();
        fixture.Customize(new AutoMoqCustomization());
        
        var dep = fixture.Freeze<Mock<IDep>>();
        dep.DefaultValue = DefaultValue.Empty;
        var myClass = fixture.Create<MyClass>();
        
        Assert.That(dep.Object.Get(5), Is.Null); //true
        Assert.That(myClass.Method(5), Is.Null); //true
    }
```

P.S. your project is amazing!

**Comment by ivansorokinf** [comment]
I managed to do it using postprocessor customization, but it seems kinda hacky and heavy for configuring integration with Moq, probably things like this should be somewhere in AutoMoq customization settings.

```
internal class MockFiller : ISpecimenCommand
{
    public void Execute(object specimen, ISpecimenContext context)
    {
        var mock = specimen as Mock;
        mock.DefaultValue = DefaultValue.Empty;
    }
}
    
internal class MockSpecification : IRequestSpecification
{
    public bool IsSatisfiedBy(object request)
    {
        var requestType = request as Type;
        if (requestType == null)
            return false;

        return typeof(Mock).IsAssignableFrom(requestType);
    }
}
```
adding to fixture
```
var specimenBuilder = new Postprocessor(new MethodInvoker(new ModestConstructorQuery()),
                                                new MockFiller());
fixture.Customizations.Add(new FilteringSpecimenBuilder(specimenBuilder, new MockSpecification()));
fixture.Customize(new AutoMoqCustomization{ConfigureMembers = true});
```

**Comment by aivascu** [comment]
@ivansorokinf unfortunately I can't tell why exactly this is the default behavior of AutoFixture, since there are no records as to why this decision was made. However knowing Mark's view of interface consistency, I think it's safe to assume that the intention was to, promote good design by making the mocks respect the interface they were implementing.

What I mean is that if a method says that it will return an instance of type `Dummy`, then there is no way it wouldn't return it. If it couldn't return the instance then it would throw. Even if this was not the reason this was originally added, I think it's a good one to leave this behavior as is.

Now to answer your question about how to replace this behavior. The part that sets this configuration is the `MockPostprocessor`, as you've already mentioned in your original post. The proper way to change this behavior is to replace it with a different implementation that would configure the `DefaultValue` to your preference.
You can either add it on top of the customization pipeline or replace the original implementation in place.

My personal suggestion is to add builders on top of the customization pipelines rather than remove, since this is faster and it's a non-destructive way to change behavior. So if you wanted you could later remove your builder and AutoFixture would return to the original behavior.

I prepared a Gist for you [here](https://gist.github.com/aivascu/eb658ff05784541cc3f130839f27e227), showing how to implement it.

**Reply by ivansorokinf** [reply]
Thanks for spending your time! I understand the logic behind creating mocks with `DefaultValue=Mock` and why it is good practice. But some use cases need 'old' behaviour. When migrating from `RhinoMocks + Autofixture` to `Moq + Autofixture`  I encountered 12k tests which are often using implicit knowledge of default for non-setupped methods and it is very expensive to rewrite them.  That is why I suggested adding some options to configure integration -- for helping big legacy projects switching from old but popular Rhino etc. Do you think it is a bad idea in general to let users customize Moq integration like this? I can try to make PR with adding option to `AutoMoqCustomization` beside `ConfigureMembers`  and 'GenerateDelegates'.

**Reply by aivascu** [reply]
@ivansorokinf it might be worth adding a configuration option to override the DefaultValue. However the new configuration should respect the existing configuration flags, and should not break the existing tests. Do you have a proposal for the API?

**Reply by ivansorokinf** [reply]
Hello, @aivascu

I see possible realization as such
1) add `DefaultValue` flag to the `AutoMoqCustomization` with `default=Default.ValueMock` if not specified
2) teach the existing `MockConfigurator<T>` to respect this value in its `ConfigureMethod`

```
private readonly DefaultValue defaultValue; // comes from constructor (AutoMoqCustomization => MockPostprocessor => this class)
private static void Configure(Mock<T> mock)
{
    mock.CallBase = ShouldUseCallBase(typeof(T));
    mock.DefaultValue = defaultValue;
}
```

It shouldn't break existing configurations as far as I see. Another realization in my opinion is adding `Action<Mock> configure` to AutoMoqCustomization settings to allow user to change not only `DefaultValue` but some other settings as `DefaultValueProvider`.  But this approach can possibly break some internal mechanisms of AutoFixture which doesn't expect these changes.
